habitat_sim/simulator.py

```diff
@@ -472,24 +448,115 @@ def __del__(self) -> None:
     def step_physics(self, dt: float, scene_id: int = 0) -> None:
         self.step_world(dt)
 
+    def draw_observation(self, sensor_reader: SensorReader) -> None:
```

Comment:
Using the SensorReader here and in `get_observation` seems wrong.  The reader is an implementation detail to cache things, not the driver.  The sensor is what drives.

Reply:
I totally agree with you. @erikwijmans

Reply:
> The sensor is what drives.
@erikwijmans : when you said this, my understanding was it should be: `sensor.draw_observation()`. Correct?

Reply:
Thanks for the insight Erik, the torch buffer type is a big concern. Based on what you've said here, it seems we will have to keep the Sensor wrapper class in python, and subsequently, keep the global Dict[str, Sensor] in simulator as well. Although it might make sense to think of the python Sensor as a SensorReader, it seems this might not be something the users would need to know.

Reply:
@erikwijmans @vauduong . I think I figured out hwo to read into a Torch / Numpy buffer directly from C++. PyBind11 has a C++ Wrapper for Python meaning you can execute Python code in the generated functions (including importing pure python modules at runtime).  https://pybind11.readthedocs.io/en/stable/advanced/pycpp/object.html#accessing-python-libraries-from-c . We could do attach the minimal pure python functions to the Sensor Class using that.